### PR TITLE
pyclass: move flags to PyClassImpl

### PR DESCRIPTION
**Pull request by davidhewitt**
This is a (mostly internal) refactoring to move some pieces of `PyTypeInfo` which aren't needed for anything other than `#[pyclass]` onto the `PyClassImpl` trait.

At the same time I cleaned up the old `FLAGS` constant which still had unneeded entries for e.g. weakref.

### CODE REVIEW

File: src/pyclass.rs [hunk 1]
```diff
@@ -153,6 +153,7 @@ pub trait PyClass:
 struct TypeSlots(Vec<ffi::PyType_Slot>);
 
 impl TypeSlots {
+    #[inline]
```
Comment (kngwyu) [inline src/pyclass.rs:156]: Is this effective?
This is not a public function so the compiler would automatically make this inline if needed.

File: src/pyclass.rs [hunk 2]
```diff
@@ -174,6 +175,7 @@ fn get_type_name<T: PyTypeInfo>(module_name: Option<&str>) -> PyResult<*mut c_ch
     })
 }
 
+#[inline]
```
Comment (kngwyu) [inline src/pyclass.rs:178]: Same as above.

File: src/class/impl_.rs [hunk 3]
```diff
@@ -31,7 +31,19 @@ impl<T> Copy for PyClassImplCollector<T> {}
 ///
 /// Users are discouraged from implementing this trait manually; it is a PyO3 implementation detail
 /// and may be changed at any time.
-pub trait PyClassImpl: Sized {
+pub trait PyClassImpl: Sized + PyTypeInfo {
```
Comment (kngwyu) [inline src/class/impl_.rs:34]: I'm not sure this reduces the complexity of our (somewhat complex) trait hierarchy...

**Review by kngwyu — COMMENTED** [review]
Thanks!
Looks like reasonable changes, but why `PyClassImpl` instead of `PyClass`?
I understand that `PyClassImpl` is for collecting methods by pseudo-specialization trick, so I don't see the necessity to place these flags under `PyClassImpl`.

File: src/pyclass.rs (hunk 1 above)
Comment (davidhewitt) [reply src/pyclass.rs:156]: I felt it couldn't hurt to add this attribute here. I can try and measure the benchmarks to see if I can get any meaningful difference.

File: src/class/impl_.rs (hunk 3 above)
Comment (davidhewitt) [reply src/class/impl_.rs:34]: I agree it's sad, this is because `PyClassAlloc` wanted `PyTypeInfo`. Maybe I can do some other adjusting.

**Comment by davidhewitt** [comment]
> Looks like reasonable changes, but why `PyClassImpl` instead of `PyClass`?

Overall I see `PyClassImpl` as the pieces of `PyClass` which we don't consider stable API (it's `#[doc(hidden)]`). So unless we are sure this is useful for users and it's not going to need to change, it should go on `PyClassImpl`.

To reduce the complex trait hierarchy I was thinking to merge `PyClassAlloc` into `PyClassImpl` also.

**Comment by kngwyu** [comment]
How about moving some flags under `PyClassAlloc`?
They are actually related only to allocation.

**Comment by davidhewitt** [comment]
> How about moving some flags under PyClassAlloc? They are actually related only to allocation.

Only the `IS_SUBCLASS` flag is used during the `default_new` implementation - the others are used in `py_class_flags` which isn't anything to do with allocation.

I'll have a go a follow-up commit to try refactoring `PyClassAlloc` a bit. 👍

File: src/pyclass.rs (hunk 1 above)
Comment (davidhewitt) [reply src/pyclass.rs]: Okay, removed!

**Comment by davidhewitt** [comment]
I've pushed a commit which removes the `#[inline]` and puts `PyTypeInfo` bound onto `PyClassAlloc` instead of `PyClassImpl`. Also renamed `T::DESCRIPTION` to `T::DOC` for simplification.

I looked at doing further work to `PyClassAlloc` in this PR but realised there are a lot of complications with that trait as-is, especially with #991.

So I will try and push something experimental in a separate PR later.

**Review by kngwyu — APPROVED** [review]
Nice, thank you!

**Pull request merged**